### ISSUE TITLE
TypeVar incompatible with nested coroutines

### ISSUE THREAD

**Issue by bushkov** [python/mypy#13408]
**Bug Report**

When a coroutine defines a nested coroutine and both of them have a return type constrained by a `TypeVar`, MyPy incorrectly complains about incompatible return value types.

**To Reproduce**

The following code results in a false positive:
```python
from typing import TypeVar

T = TypeVar("T", int, str)

async def f(x: T) -> T:
    async def g() -> T:
        return x
    return await g()
```

Note that if the nested coroutine is changed to a function, MyPy gives no errors:
```python
from typing import TypeVar

T = TypeVar("T", int, str)

async def f(x: T) -> T:
    def g() -> T:
        return x
    return g()
```

**Expected Behavior**

No errors.

**Actual Behavior**

```
main.py:7: error: Incompatible return value type (got "int", expected "Coroutine[Any, Any, int]")
main.py:7: error: Incompatible return value type (got "str", expected "Coroutine[Any, Any, str]")
```

**Your Environment**

- Mypy version used: 0.971
- Mypy command-line flags: none
- Mypy configuration options from `mypy.ini` (and other config files): none
- Python version used: 3.10.6
- Operating system and version: Ubuntu 22.04.1

**Comment by jonathanberthias** [comment]
@AlexWaygood This was fixed in [1.4.0](https://github.com/python/mypy/releases/tag/v1.4.0), the given example [doesn't error anymore](https://mypy-play.net/?mypy=1.4.0&python=3.11&gist=ad160eadbd99da62c8fb62d88adb894c).
